### ISSUE TITLE
Error when the num of buildings does not match the num of nodes

### ISSUE THREAD

**Issue by taniguchitakara** [jo2lxq/wafl#18]
<summary>
label: 0 train_data: 98 test_data: 25
label: 1 train_data: 125 test_data: 32
label: 2 train_data: 101 test_data: 26
label: 3 train_data: 88 test_data: 21
label: 4 train_data: 83 test_data: 20
label: 5 train_data: 93 test_data: 24
label: 6 train_data: 103 test_data: 26
label: 7 train_data: 101 test_data: 26
label: 8 train_data: 104 test_data: 26
label: 9 train_data: 88 test_data: 22
label: 10 train_data: 120 test_data: 31
label: 11 train_data: 86 test_data: 22
the number of data for training 0-th node: 119
the number of data for training 1-th node: 119
the number of data for training 2-th node: 119
the number of data for training 3-th node: 119
the number of data for training 4-th node: 119
the number of data for training 5-th node: 119
the number of data for training 6-th node: 119
the number of data for training 7-th node: 119
the number of data for training 8-th node: 119
the number of data for training 9-th node: 119
batch_size: 16
contact file: rwp_n10_a0500_r100_p10_s01.json
do not use filter
test transform: Compose(
    CenterCrop(size=(224, 224))
    ConvertImageDtype()
    Normalize(mean=(0.4739702343940735, 0.4647347033023834, 0.43512433767318726), std=(0.24572011828422546, 0.24565722048282623, 0.25671228766441345))
)
train transform: Compose(
    RandomResizedCrop(size=(224, 224), scale=(0.4, 1.0), ratio=(0.75, 1.3333), interpolation=bilinear, antialias=warn)
    ConvertImageDtype()
    Normalize(mean=(0.4725642502307892, 0.4632246494293213, 0.4340682625770569), std=(0.2503652274608612, 0.24852073192596436, 0.26192405819892883))
    RandomErasing(p=0.5, scale=(0.02, 0.33), ratio=(0.3, 3.3), value=0, inplace=False)
)
optimizer: SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    lr: 0.01
    maximize: False
    momentum: 0.9
    nesterov: False
    weight_decay: 0
)
pre-training optimizer: SGD (
Parameter Group 0
    dampening: 0
    differentiable: False
    foreach: None
    lr: 0.01
    maximize: False
    momentum: 0.9
    nesterov: False
    weight_decay: 0
)
net:
 ===============================================================================================
Layer (type:depth-idx)                        Output Shape              Param #
===============================================================================================
VisionTransformer                             [1, 12]                   768
├─Conv2d: 1-1                                 [1, 768, 14, 14]          (590,592)
├─Encoder: 1-2                                [1, 197, 768]             151,296
│    └─Dropout: 2-1                           [1, 197, 768]             --
│    └─Sequential: 2-2                        [1, 197, 768]             --
│    │    └─EncoderBlock: 3-1                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-2                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-3                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-4                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-5                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-6                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-7                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-8                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-9                 [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-10                [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-11                [1, 197, 768]             (7,087,872)
│    │    └─EncoderBlock: 3-12                [1, 197, 768]             (7,087,872)
│    └─LayerNorm: 2-3                         [1, 197, 768]             (1,536)
├─Sequential: 1-3                             [1, 12]                   --
│    └─Linear: 2-4                            [1, 256]                  196,864
│    └─ReLU: 2-5                              [1, 256]                  --
│    └─Linear: 2-6                            [1, 12]                   3,084
===============================================================================================
Total params: 85,998,604
Trainable params: 199,948
Non-trainable params: 85,798,656
Total mult-adds (Units.MEGABYTES): 172.66
===============================================================================================
Input size (MB): 0.60
Forward/backward pass size (MB): 104.09
Params size (MB): 229.99
Estimated Total Size (MB): 334.68
===============================================================================================
<summary>

When I conducted experiments when the num of buildings does not match the num of nodes, it seems that the error happens in the dataloader from my perspective (Please check the attached image)

The attached image is the result when num_nodes = 10 and num_buildings = 12. There is a "slip" in the confusion matrix. In my opinion, model works correctly but somehow the error happens when training. In this confusion matrix, 10,11 seems to be trained as 2,3

![Image](https://github.com/user-attachments/assets/2f8de6d3-7fd6-483b-8038-843ff5cb260b)

**Comment by taniguchitakara** [comment]
I found it is because some hard-coding. For example 

[https://github.com/jo2lxq/wafl/blob/d0a93f683df2463c63f0944788af214e81038029/WAFL-ViT/src/functions/model_exchange.py#L157](url)

In this code,
`
def update_nets_vit(net, contact, fl_coefficient):
    local_model = [{} for _ in range(10)]
    recv_models = [[] for _ in range(10)]
    for n in range(10):
        local_model[n] = net[n].heads.state_dict()
        nbr = contact[str(n)]  # the nodes n-th node contacted
        recv_models[n] = []
        for k in nbr:
            recv_models[n].append(
                copy.deepcopy(net[k].heads.state_dict())
            )  # create a new object to avoid modifying the original object
            # recv_models[n].append(local_model[k])

    # mixture of models
    for n in range(10):
        update_model = copy.deepcopy(
            recv_models[n]
        )  # create a new object to avoid modifying the original object
        # update_model = recv_models[n]
        n_nbr = len(update_model)  # how many nodes n-th node contacted

        # put difference of n-th node models and k-th contacted node into update_model[k]
        for k in range(n_nbr):
            for key in update_model[k]:
                update_model[k][key] = recv_models[n][k][key] - local_model[n][key]

        # mix to local model
        for k in range(n_nbr):
            for key in update_model[k]:
                local_model[n][key] += (
                    update_model[k][key] * fl_coefficient / float(n_nbr + 1)
                )
    # update nets
    for n in range(10):
        nbr = contact[str(n)]
        if len(nbr) > 0:
            net[n].heads.load_state_dict(local_model[n])
`

These

```
    local_model = [{} for _ in range(10)]
    recv_models = [[] for _ in range(10)]
```

are specified for 10 nodes. Other than this , I am now looking for and fixing to apply for the custom dataset. 

So I would like you to inform me about these hard-coding in your memory. (I would like to collaborate with you and fix them if you have time)

Regards

**Comment by jo2lxq** [comment]
Thank you for posting this issue. I will consult amura870 and ueda0913.

**Comment by ueda0913** [comment]
Thank you for pointing out this problem. 
To begin with, we need to modify hard-cordings as you mentioned because they deteriorate model training. We will address this issue. However, the errors in the confusion matrix are likely caused not by this hard-cording, but the implementation of `torchvision.datasets.ImageFolder`.
`ImageFolder` automatically sets label indices based on the directory names in ascending order as strings. Therefore, if you use directories named "10" and "11", they will be set between "1" and "2". This leads models to train label 10 as 2, label 11 as 3, and labels 2-9 as labels 4-11. 
We will need more time to verify this solution before merging it into the main branch. Please allow us some time to conduct further testing.

**Comment by jo2lxq** [comment]
I see. Does this mean the folder names should start with 00, 01, 02, ..., up to 99, if there are more than 10 labels?

**Comment by ueda0913** [comment]
Yes. But, that is inconvenient, so I am trying not to use `ImageFolder` itself.

**Comment by taniguchitakara** [comment]
Thank you! 
I checked it worked correctly! Thanks!
